CIPs/cip-draft_poprf.md

```diff
@@ -0,0 +1,474 @@
+  name: "ODIS Linear Backoff Domain"
+  version: 1
+  cap: 10
+  salt: "saltvalue"
+}
+```
+
+<!-- TODO(victor) Verify that there are no mistakes in this example. -->
+```text
+domainSeparator = keccak256(
+  keccak256("EIP712Domain(string name,string version)") ||
+  keccak256("ODIS Linear Backoff Domain") || keccak256("1")
+)
+typeHash = keccak256(
+  "LinearBackoffDomain(int256 cap, Optional<int256> refresh, Optional<string> salt)" ||
+  "Optional<int256>(bool defined,int256 value)" ||
+  "Optional<string>(bool defined,string value)"
+)
+structHash = keccak256(
+  typeHash ||
+  0x000000000000000000000000000000000000000000000000000000000000000A ||
+  keccak256(
+    keccak256("Optional<int256>(bool defined,int256 value)") ||
+    0x0000000000000000000000000000000000000000000000000000000000000000 ||
+    0x0000000000000000000000000000000000000000000000000000000000000000
+  ) ||
+  keccak256(
+    keccak256("Optional<string>(bool defined,string value)") ||
+    0x0000000000000000000000000000000000000000000000000000000000000001 ||
+    keccak256("saltvalue")
+  )
+)
+encoded = 0x1901 || domainSeparator || structHash
+```
+
+### Signatures
+
+Signatures are a useful primitive for authenticated rate limiting (e.g. allowing a higher rate limit
+to the owners of verified accounts on Celo). In order to facilitate this, the entire request struct
+should be deterministically serializeable, specifically using EIP-712.
+
+Domains that make use of signatures for authentication of requests (e.g. the `PepperedDomain`
+example below) should include a field in the domain options for specification of a signature.
+Additional fields may also be added, such as a nonce to prevent replaying requests.
+
+Signatures should be verified over the entire request, excluding the signature itself. Specifically,
+the message against which a signature is verified should be obtained by first removing the signature
+from the full `GetDomainRestrictedSigRequest` object, then serializing it via EIP-712.
+
+Serializing the request with EIP-712 should use the domain `{ name: "ODIS Domain Restricted
+Signature Request", version: "1" }` and type defined by applying the type mapping [defined
+above](#mapping-typescript-to-eip-712-types) to the `GetDomainRestrictedSigRequest` type.
+
+### HTTP Transport
+
+When requested over HTTP, which is the primary proposed transport, the following HTTP semantics
+SHOULD be followed.
+
+If a given `name` and `version` are not recognized by an ODIS signer, it SHOULD return a `501: Not
+Implemented` error. If a given `name` and `version` are recognized, but are deprecated, it SHOULD
+return a `405: Method Not Allowed` error. Signers SHOULD keep a list of all old domain names and
+version, even if they are no longer supported.
+
+When a user's request is rejected for violating the rate limiting of a domain, the service SHOULD
+respond with `429: Too Many Requests`. If the waiting would allow for additional requests against
```

Comment:
I think we currently use 403 for this but maybe 429 is more appropriate

Reply:
I think so, especially because the rate limit may have a time component which can be indicated with `Retry-After`.